src/sage/graphs/graph_decompositions/slice_decomposition.pyx

```diff
@@ -0,0 +1,970 @@
+# cython: binding=True
+# distutils: language = c++
+# distutils: extra_compile_args = -std=c++11
+r"""
+Slice decomposition
+
+This module implements an extended lexBFS algorithm for computing the slice
+decomposition of undirected graphs and the class :class:`~SliceDecomposition` to
+represent such decompositions.
+
+AUTHORS:
+
+- Cyril Bouvier (2024-06-25): initial version
+"""
+# ****************************************************************************
+# Copyright (C) 2024 Cyril Bouvier <cyril.bouvier@lirmm.fr>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#                  https://www.gnu.org/licenses/
+# ****************************************************************************
+
+from libcpp.algorithm cimport swap
+from cython.operator cimport dereference as deref
+
+from sage.graphs.base.c_graph cimport CGraphBackend
+from sage.data_structures.bitset_base cimport bitset_in
+
+cdef void extended_lex_BFS(
+        CGraph cg, vector[int] &sigma, vector[int] *sigma_inv,
+        int initial_v_int, vector[int] *pred, vector[size_t] *xslice_len,
+        vector[vector[int]] *lex_label) except *:
+    r"""
+    Perform a extended lexicographic breadth first search (LexBFS) on the
+    undirected graph `G`. In addition to computing a LexBFS ordering, the
+    extended LexBFS algorithm can be used to compute the slice decomposition of
+    the graph.
+
+    This function implements the `O(n+m)` time algorithm proposed in [HMPV2000]_
+    and [TCHP2008]_.
+
+    INPUT:
+
+    - ``cg`` -- a ``CGraph``. This function ignores loops and multiple edges and
+      assumes that the graph is undirected.
+
+    - ``sigma`` -- vector of ``int`` to store the ordering of the vertices
+      resulting from the LexBFS traversal. At the end, the vector will have size
+      `n` (the number of vertices of the graph).
+
+    - ``sigma_inv`` -- a pointer to a vector to store the inverse of the
+      permutation ``sigma``. ``sigma_inv`` can be ``NULL`` if the caller does
+      not need it (but, note that, the inverse of ``sigma`` is still needed by
+      the algorithm, so it does not save time nor memory to have ``sigma_inv``
+      equal to ``NULL``). At the end, if ``sigma_inv`` is not NULL, the vector
+      pointer by it will have size `n` (the number of vertices of the graph)
+      and will satisfy:
+        * sigma[deref(sigma_inv)[v_int]] = v_int
+        * deref(sigma_inv)[sigma[i]] = i
+
+    - ``initial_v_int`` -- the first vertex to consider, can be -1.
```

Comment:
you may explain the meaning of -1

Reply:
done in 2d5002fdfd6ed31286fcd8dc260c742ee408e90e